[Cookie Store API] Ensure subdomains can set/get cookies for higher-level domains

RupinMittal opened the pull request:
#### 4413e221bc3419c996bc44b737d6af540908a0ca
<pre>
[Cookie Store API] Ensure subdomains can set/get cookies for higher-level domains
<a href="https://bugs.webkit.org/show_bug.cgi?id=285890">https://bugs.webkit.org/show_bug.cgi?id=285890</a>
<a href="https://rdar.apple.com/142427907">rdar://142427907</a>

Reviewed by Sihui Liu.

Step 9 in the Cookie Store API spec for setting a cookie (<a href="https://wicg.github.io/cookie-store/#set-a-cookie)">https://wicg.github.io/cookie-store/#set-a-cookie)</a>
says that if a non-null domain is passed in, then:

1. If domain starts with U+002E (.), then return failure.
2. If host does not equal domain and host does not end with U+002E (.)
   followed by domain, then return failure.

This means that on the test website (which has a host of static-safari.apple.com),
setting a cookie with the domain &quot;apple.com&quot; should work.

But currently, it does not.

More generally speaking, this means subdomains should be able to set/get cookies for higher level domains,
but currently are unable to.

The issue is that we&apos;re not following all the steps of setting a cookie. After the pre-processing steps
in the #set-a-cookie part of the spec linked above, the spec says to follow the steps here:
<a href="https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-14#name-storage-model">https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-14#name-storage-model</a>

Step 10 of this says:
If the domain-attribute is non-empty:
    If the canonicalized request-host does not domain-match the domain-attribute:
        Abort these steps and ignore the cookie entirely.
    Otherwise:
        Set the cookie&apos;s host-only-flag to false.
        Set the cookie&apos;s domain to the domain-attribute.
Otherwise:
    Set the cookie&apos;s host-only-flag to true.
    Set the cookie&apos;s domain to the canonicalized request-host.

A host-only cookie is one accessible only by the exact host/domain it was set by. (If you don&apos;t
specify a domain, the host will be used as the domain, so they&apos;ll be same).

The way host-only only works is a CFNetwork quirk and is explained in a comment in the
normalizeCookieProperties function in CFHTTPCookie.mm:
- If domain begins with a &apos;.&apos;, host-only is false, and so the cookie is accessible by subdomains
- If not, it&apos;s considered a host-only cookie (not accessible by subdomains).

Currently, we&apos;re setting all cookies without the &apos;.&apos; (so they are host-only). This patch
updates CookieStore::set to follow step 10. If the user passes in a non-null domain which
domain-matches the host, we will prepend a &apos;.&apos; to the domain (which sets host-only to false).

This allows subdomains to set/get cookies for higher level domains.

As noted at the bottom of the spec&apos;s section about querying cookies:
<a href="https://wicg.github.io/cookie-store/#query-cookies-algorithm">https://wicg.github.io/cookie-store/#query-cookies-algorithm</a>, the host-only property is not
exposed to script. And CFNetwork&apos;s quirk with the &apos;.&apos; should not be exposed to script either.
So this &apos;.&apos; is stripped when the cookie is returned by any getter of the Cookie Store API. How: get
operations always return a CookieListItem constructed from the Cookie. So upon construction from a
Cookie, we force CookieListItem to strip the beginning &apos;.&apos; from the domain if if exists.

Testing:
- Checking that subdomains can set/get cookies for higher-level domains (checking host-only is set
to false) is done by the new API test CookieStoreSetCookieForHigherLevelDomain in CookieStoreAPI.mm

- Checking that returning a cookie will not have a prepended &apos;.&apos; is already done by existing layout tests.
One of the tests in this file below sets a cookie by specifying a domain and checks the returned domain:
<a href="https://github.com/web-platform-tests/wpt/blob/master/cookie-store/cookieListItem_attributes.https.any.js">https://github.com/web-platform-tests/wpt/blob/master/cookie-store/cookieListItem_attributes.https.any.js</a>

Changes to WPT layout tests:
There is a test in cookieStore_delete_arguments.https.any.js that sets a cookie, gets the cookie, and
passes the result of the get into delete. This test was passing before this change but now fails to
delete the cookie. The issue is that the result of the get is a CookieListItem (which contains a
domain). When we set the cookie, we don&apos;t specify a domain, so the domain is set to &quot;localhost&quot;.
The returned result (which contains &quot;localhost&quot;) is passed to delete. Since the delete call recieves
a passed-in domain, it adds the &apos;.&apos; and tries to delete the cookie with domain &quot;.localhost&quot;. Of course,
there is no such cookie and so our original cookie doesn&apos;t get deleted.

But CookieListItem shouldn&apos;t contain domain at all--we don&apos;t want to expose this property. CookieListItem
should contain only name and value. So we change the test to rely only on the properties we actually
expose to script--so the test continues to pass now. Similar change in the other changed test. We&apos;ll
upstream these changes to WPT soon.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any.js:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.js:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.serviceworker-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
(WebCore::CookieListItem::CookieListItem):
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CookieStoreAPI.mm: Added.
(TestWebKitAPI::TEST(WebKit, CookieStoreSetCookieForHigherLevelDomain)):

Canonical link: <a href="https://commits.webkit.org/288983@main">https://commits.webkit.org/288983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b66fad1c461210e3c09f8b72e709a5910d1df38e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84959 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39355 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/90108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/36015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87044 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4779 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12664 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/90108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/36015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88004 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/77207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/90108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/3525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/31434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/35088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/32242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91481 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12303 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12532 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/73017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/73716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/18105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/16556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13239 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12251 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/12087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/15583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13831 "Failed to checkout and rebase branch from PR 38972") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -442,6 +442,15 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+        // the Cookie Store API. The above checks only allow setting a cookie for one's own domain or a higher level domain.
+        // So if the host does not equal the domain, it means a subdomain is setting the cookie for a higher-level domain.
+        if (host != cookie.domain)
+            cookie.domain = makeString('.', cookie.domain);
```
I think the details could go to commit message; it could be something simple like:

> // In CFNetwork, domain without leading dot means host-only cookie.
> // Prepend dot to domain to make the cookie accessible by subdomains.

Also, why do we only prepend the dot when domain does not equal to host?

RupinMittal replied:
Sounds good--I'll update the comment when I re-upload. 

> Also, why do we only prepend the dot when domain does not equal to host?

This is a good question. Initially, when I tested Chrome, it looked like it was only prepending the dot for the higher level domains. But actually now that I look again, this is the behavior when on the test site (static-safari.apple.com):

no domain specified --> then look at the domain in web inspector, domain is "static-safari.apple.com" (WITHOUT a dot). 

specify a domain that is equal to host (static-safari.apple.com in this case) --> then look at the domain in web inspector, domain is ".static-safari.apple.com" (WITH a dot). 

specify a higher-level domain "apple.com" --> then look at the domain in web inspector, domain is ".apple.com" (WITH a dot).

So if we want to match Chrome, the logic should be: prepend a dot only if a domain was passed in (regardless of if the domain is higher level or the same as host). 

I'll update this in the next iteration.

RupinMittal replied:
Or should we prepend the dot in all cases?

RupinMittal replied:
Actually, step 10 of the spec says exactly these steps I described above: https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-14#name-storage-model. I'll make the changes..

RupinMittal replied:
Re-uploaded with the change and a test!

szewai commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -425,15 +425,15 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
     cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
 
-    cookie.domain = options.domain.isNull() ? domain : WTFMove(options.domain);
+    cookie.domain = options.domain.isNull() ? domain : options.domain;
```
I think you don't need to set cookie.domain if options.domain is null. If domain is not set, cookie is host-only (so no need to specify domain):
```
cookie.domain = options.domain;
if (!cookie.domain.isNull()) {
    ....
    // In CFNetwork, a domain without a leading dot means host-only cookie.
    // If a non-null domain was passed in, prepend dot to domain to set
    // host-only to false and make the cookie accessible by subdomains.
    cookie.domain = makeString('.', cookie.domain);
}

```

RupinMittal replied:
If we don't set cookie.domain, it will be null.

But the spec says that the if no domain is passed in, it should default to the current document/service worker's domain: https://wicg.github.io/cookie-store/#CookieStore-set

Plus, CFNetwork doesn't let us set cookies with null domain anyways.

szewai replied:
I see; yes then let's do the original way.

RupinMittal committed:
[Cookie Store API] Ensure subdomains can set/get cookies for higher-level domains
https://bugs.webkit.org/show_bug.cgi?id=285890
rdar://142427907

Reviewed by Sihui Liu.

Step 9 in the Cookie Store API spec for setting a cookie (https://wicg.github.io/cookie-store/#set-a-cookie)
says that if a non-null domain is passed in, then:

1. If domain starts with U+002E (.), then return failure.
2. If host does not equal domain and host does not end with U+002E (.)
   followed by domain, then return failure.

This means that on the test website (which has a host of static-safari.apple.com),
setting a cookie with the domain "apple.com" should work.

But currently, it does not.

More generally speaking, this means subdomains should be able to set/get cookies for higher level domains,
but currently are unable to.

The issue is that we're not following all the steps of setting a cookie. After the pre-processing steps
in the #set-a-cookie part of the spec linked above, the spec says to follow the steps here:
https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-14#name-storage-model

Step 10 of this says:
If the domain-attribute is non-empty:
    If the canonicalized request-host does not domain-match the domain-attribute:
        Abort these steps and ignore the cookie entirely.
    Otherwise:
        Set the cookie's host-only-flag to false.
        Set the cookie's domain to the domain-attribute.
Otherwise:
    Set the cookie's host-only-flag to true.
    Set the cookie's domain to the canonicalized request-host.

A host-only cookie is one accessible only by the exact host/domain it was set by. (If you don't
specify a domain, the host will be used as the domain, so they'll be same).

The way host-only only works is a CFNetwork quirk and is explained in a comment in the
normalizeCookieProperties function in CFHTTPCookie.mm:
- If domain begins with a '.', host-only is false, and so the cookie is accessible by subdomains
- If not, it's considered a host-only cookie (not accessible by subdomains).

Currently, we're setting all cookies without the '.' (so they are host-only). This patch
updates CookieStore::set to follow step 10. If the user passes in a non-null domain which
domain-matches the host, we will prepend a '.' to the domain (which sets host-only to false).

This allows subdomains to set/get cookies for higher level domains.

As noted at the bottom of the spec's section about querying cookies:
https://wicg.github.io/cookie-store/#query-cookies-algorithm, the host-only property is not
exposed to script. And CFNetwork's quirk with the '.' should not be exposed to script either.
So this '.' is stripped when the cookie is returned by any getter of the Cookie Store API. How: get
operations always return a CookieListItem constructed from the Cookie. So upon construction from a
Cookie, we force CookieListItem to strip the beginning '.' from the domain if if exists.

Testing:
- Checking that subdomains can set/get cookies for higher-level domains (checking host-only is set
to false) is done by the new API test CookieStoreSetCookieForHigherLevelDomain in CookieStoreAPI.mm

- Checking that returning a cookie will not have a prepended '.' is already done by existing layout tests.
One of the tests in this file below sets a cookie by specifying a domain and checks the returned domain:
https://github.com/web-platform-tests/wpt/blob/master/cookie-store/cookieListItem_attributes.https.any.js

Changes to WPT layout tests:
There is a test in cookieStore_delete_arguments.https.any.js that sets a cookie, gets the cookie, and
passes the result of the get into delete. This test was passing before this change but now fails to
delete the cookie. The issue is that the result of the get is a CookieListItem (which contains a
domain). When we set the cookie, we don't specify a domain, so the domain is set to "localhost".
The returned result (which contains "localhost") is passed to delete. Since the delete call recieves
a passed-in domain, it adds the '.' and tries to delete the cookie with domain ".localhost". Of course,
there is no such cookie and so our original cookie doesn't get deleted.

But CookieListItem shouldn't contain domain at all--we don't want to expose this property. CookieListItem
should contain only name and value. So we change the test to rely only on the properties we actually
expose to script--so the test continues to pass now. Similar change in the other changed test. We'll
upstream these changes to WPT soon.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any.js:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.js:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.serviceworker-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
(WebCore::CookieListItem::CookieListItem):
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CookieStoreAPI.mm: Added.
(TestWebKitAPI::TEST(WebKit, CookieStoreSetCookieForHigherLevelDomain)):

Canonical link: https://commits.webkit.org/288983@main
